ENH: Added exact computation for Kolmogorov-Smirnov two-sample test.

pvanmulbregt opened the pull request:
Replaced approximation with exact computation for two-sided test `stats.ks_2samp`. 
Added a one-sided, two-sample KS test. 
Added a keyword `alternative` to `stats.ks_2samp`. so the signature becomes
```
stats.ks_2samp(data1, data2, alternative='two-sided')
```
Slightly increase accuracy of the test statistic.
Added test cases.
Updated the stats tutorial with slightly changed probabilities.

**WARNING: This PR changes returned results.**.  
An extreme example occurs when comparing two samples of length 1, `[X0]` and `[Y0]`  
The computed test statistic is `1.0`, and the returned probability changes from `0.289` to `1.0`(!) 
[0.289 came from using the Smirnov approximation suitable for large `n`, even though `n=1` and is not large.  
Then probability 1.0 arises because when comparing two size 1 samples, the smallest Empirical CDF gap is 1.0 (as the distributions are continuous so no common values), hence `Pr(Dn,m >= 1.0) = 1`. So the new exactly computed value is much more accurate for n small and moderate sized n.

This is the second of 3 planned PRs for (the first is gh-9719).  The 3rd will be a smaller one switching which functions are called inside `stats.{kstest/ks_2samp}` with an indication of what results may change.

pvanmulbregt commented:
Added an Issue gh-9786 to document some inaccuracies with the `ks_2sample`. This PR fixes them

chrisb83 commented:
There is also issue #8977 (inconsistency between stats and mstats)

pvanmulbregt commented:
I'm pushing off dealing with the consistency concerns in #8977 and #8748 until `kstwo` (#9719) and the updated `ks_2samp` (#9753) are both available in the same build.  That will be part of the so-called "PR3", the 3rd PR in the trilogy.  I created #9797 to keep track of the myriad correspondences, both in my mind and reviewers...

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4840,13 +5052,19 @@ def ks_2samp(data1, data2):
 
-    This is the two-sided test, one-sided tests are not implemented.
-    The test uses the two-sided asymptotic Kolmogorov-Smirnov distribution.
+    For small enough sample sizes, the computation is exact.  For large sizes,
```
Maybe add what small enough means

pvanmulbregt replied:
I was deliberately vague here, as I didn't want to commit to any particular size in the API's documentation and be bound to that forever. Computational resources have changed greatly since the first version of the K-S testing code went into SciPy.
Currently I'm using `max(n1, n2) <= 10,000` to be small enough.
Suggestions for how to word this?

chrisb83 replied:
i think you can write is explicitly, even if one decides to change the default value later, i don't see a problem to simply update the doc.
This is what the R documentation of `ks.test` says:
"Exact p-values are not available for the two-sample case if one-sided or in the presence of ties. If exact = NULL (the default), an exact p-value is computed if the sample size is less than 100 in the one-sample case and there are no ties, and if the product of the sample sizes is less than 10000 in the two-sample case."

pvanmulbregt replied:
Changed.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4814,18 +4819,225 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    alternative : {'two-sided', 'less','greater'}, optional
+        Defines the alternative hypothesis (see explanation above).
+        Default is 'two-sided'.
+    mode : 'auto' (default), 'exact' or 'asymp', optional
```
The descriptions of alternative and mode are Not aligned

pvanmulbregt replied:
Fixed.

chrisb83 replied:
i think it should be mode : {'auto', ….}

pvanmulbregt replied:
Changed.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4840,13 +5052,19 @@ def ks_2samp(data1, data2):
     reject the hypothesis that the distributions of the two samples
     are the same.
 
+    In the one-sided test, the alternative is that the empirical
```
It is confusing to associate data1 and G and data2 and F

pvanmulbregt replied:
I agree.  
It is this way to match the description (and arguments) of `kstest(rvs, cdf,...)`, which uses `G` as the ECDF of the observed random variable, and `F` for the given distribution.
[In the 3rd PR of this K-S series, `kstest` and `ks_2samp` are unified, so that 
in `kstest(rvs, cdf,...)`, `rvs` can be either data (aka `data1`), or a callable; and `cdf` can be either a CDF or data (aka `data2`).]

Suggestions for how the doc should be written for these two functions now?

chrisb83 replied:
I think if `(data1, F)` and `(data2, G)` would be ok (this ordering is more natural in my view, i.e. 1 <= 2, F <= G) or `F_1` and `F_2` (probably does not look nice)

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4881,20 +5099,78 @@ def ks_2samp(data1, data2):
     n1 = data1.shape[0]
     n2 = data2.shape[0]
+    if min(n1, n2) == 0:
+        return Ks_2sampResult(1.0, 1.0)
```
Does this make sense? Or raise an Error?

pvanmulbregt replied:
Changed to raise a ValueError.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4881,20 +5099,78 @@ def ks_2samp(data1, data2):
+    elif mode == 'exact':
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        if n1g >= np.iinfo(np.int).max / n2g:
+            mode = 'asymp'
```
One should add a warning in that case

pvanmulbregt replied:
Warning added.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4881,20 +5099,78 @@ def ks_2samp(data1, data2):
+                    prob = _compute_prob_outside_square(n1, h)
+                else:
+                    prob = 1 - _compute_prob_inside_method(n1, n2, g, h)
+                    print('prob=', prob)
```
Probably a line added during development :)

pvanmulbregt replied:
Yep. Now removed.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -4881,20 +5099,78 @@ def ks_2samp(data1, data2):
+                        prob = num_paths / bin
+                    except FloatingPointError:
+                        pass
+    if not np.isfinite(prob) or prob > 1 or prob < 0:
```
should one warn the user that s.th. went wrong?

pvanmulbregt replied:
This branch is entered if either a) mode !='exact', or b) mode=='exact' and the computation in `_count_paths_outside_method()` overflowed; or c) mode=='exact' and or `_count_paths_outside_method` incurred roundoff error putting the result outside of the interval `[0,1]`.
(`_compute_prob_inside_method()` and `_compute_prob_outside_square()` can be done without overflowing by always working with probs, but `_count_paths_outside_method` has to compute the number of legal paths, and this can involve very large binomial coefficients.)
In case a) there is no error, as no computation has been done.
If the sample sizes are large enough to cause case b) or c), then switching to `asymp` should give a decent approximation to the true probability.  I've reworked to make clearer and raise a warning in cases b) and c) if the passed in value of mode is 'exact'.

chrisb83 replied:
makes sense, thanks for the clarification

pvanmulbregt commented:
Rebased and force-pushed to get past pypy3 CI problem #9837, #9840.

pvanmulbregt commented:
@chrisb83 Do the changes resolve your comments?  Are there additional changes needed?

chrisb83 commented:
Yes, my comments have been addressed. I will review the calcuation and testing soon.

chrisb83 commented:
Question on the test cases: where do the expected values come from?

pvanmulbregt commented:
> Question on the test cases: where do the expected values come from?

The small length test cases were done by hand, following Hodges. The larger ones are the output of running the code.

`Hodges, J.L. Jr.,  "The Significance Probability of the Smirnov Two-Sample Test," Arkiv fiur Matematik, 3, No. 43 (1958), 469-86.`

chrisb83 commented:
ok, i will run the test cases through R which can cmopute exact p-values if alternative is two-sided.

One comment: Isn't there a problem if there are common values in x and y? Because if one sorts them in a joint array, it is not clear if the x or the y value comes first which determines if you go up or right first in the method presented in Hodges' paper. 

According to the R documentation, the exact computation does not work if there are ties even in one sample only (I don't see that this is an issue in the construction of the path in Hodges' paper). Of course, if the null hypothesis is true, ties have a probability of zero, i.e. this could be a sign that the test should not be used (continuous distribution assumption)

chrisb83 commented:
Hi,

I ran the tests in R and attached the results as an HTML in the zip-file (cannot attach html directly). 

The good news:
- the test statistics are the same in R and SciPy
- the exact p-values are the same whenever the alternative is two-sided and there are no ties
- the approximation of one-sided p-values in R and the exact ones in SciPy show good alignment; same for the approximation of two-sided p-values if there are ties and R defaults to the approximation

Some other points:
- in the tests `MiddlingBoth` and `MediumBoth`, the p-values for alternative `greater` differ a lot when the asymptotic approximation is used. Your implementation uses eq 5.3 in Hodges' paper which comes with a disclaimer ("On the basis of a limited numerical investigation, it appears to work reasonably well in other cases and can perhaps be recommended as a general interpolation formula"). I don't know if other formulas have been proposed after 1958 paper of Hodges. Maybe this discrepancy is worth investigating? We should also clearly state which approximation is used in the docstring.
- I would argue that the test cases should not have ties in the samples or maybe only in specific cases (see my comment above). Let's discuss the treatment of ties in general, R raises a warning in that case.

[testing.zip](https://github.com/scipy/scipy/files/3053349/testing.zip)

pvanmulbregt commented:
Thanks @chrisb83 for doing this.  The case of ties is interesting.  Ties within a sample do not matter, but ties between samples would correspond to some kind of a diagonal path in Hodges' paper and  these paths are not counted at all, and it's not clear what non-zero weight to give them if they were.
`np.searchsorted` does generate the correct statistic.  If there are ties, one could adjust one of the two samples up by epsilon and compute the statistic/p-value and then adjust down by epsilon and compute the statistic/p-value, and use one of the two values (the one with the most conservative implications?)  But that would be a task for the user.
I haven't yet dug deep enough into the R implementation to see why the p-values differ in some of the cases. R uses the exact computation of K(n, d) from Marsaglia, Tsang and Wang rather than the asymptotic Kolmogorov distribution which will introduce differences. (gh-9719 adds `kwsto` to SciPy and the 3rd PR in this series uses in the two-sample case.) It is also clear from Hodges' work that using the common `N=sqrt(n*m/(n+m))` is only an approximation and that there can be large oscillations. Comments in R's `ks.test` refers to a paper of Kim and Jennrich from 1970.  But the block is also enclosed in "<FIXME>" so it is not clear the current status of the implementation.

chrisb83 commented:
I think R does not use exact p-values in the one-sided case (at least that is what the documentation says: https://stat.ethz.ch/R-manual/R-devel/library/stats/html/ks.test.html) and I guess they use `N=sqrt(n*m/(n+m))` which is the common approximation, so maybe that already explains the differences which could be checked without too much effort. I was just surprises by the magnitude of the deviation in testMediumBoth and testMiddlingBoth for the alternative `less`. 

When R uses Marsaglia, Tsang and Wang, the exact p-values coincide with the ones you compute (except a small discrepancy in testLarge, but that could be some numerical error).

Some other references I found but I don't have access to any of them:
- https://www.tandfonline.com/doi/abs/10.1080/01621459.1969.10501082
- PJ Kim, RI Jennrich: Tables of the exact sampling distribution of the two-sample Kolmogorov-Smirnov criterion, Dmn, m≤ n, Selected tables in mathematical statistics, 1973
- https://amstat.tandfonline.com/doi/abs/10.1080/01621459.1976.10481562

chrisb83 commented:
At least for the asymptotic approximation, the solution is easy: there is an error in your implementation I think

```
def asymp_check(d, n1, n2):
    m, n = max(n1, n2), min(n1, n2)
    z = np.sqrt(m*n/(m+n)) * d
    expt = -2 * z**2 - 2 * z * (m + 2*n)/np.sqrt(m*n*(m+n))/3.0 # this is 5.3 in Hodges
    # current implementation: expt = -2 * z**2 - 2 * z * (m + 2*n)/(m+n)/3.0
    prob = np.exp(expt)
    return {'R': np.exp(-2 * z**2), 'Hodges': prob}
```

Now the values are really close to the ones I get in R for the test cases where the difference was large before:
```
asymp_check(0.006, 1000, 1100) # {'R': 0.9629880410296024, 'Hodges': 0.9573185808092622}
asymp_check(0.000909, 1000, 1100) #  {'R': 0.9629880410296024, 'Hodges': 0.9573185808092622}
asymp_check(0.0066666, 500, 600) # {'R': 0.9760495363535895, 'Hodges': 0.9697601353441921}
asymp_check(0.00166666, 500, 600) # {'R': 0.9984860078502431, 'Hodges': 0.9968736028442297}
```

pvanmulbregt commented:
Thanks @chrisb83 for the careful reading.  
Fixed the implementation of Hodges' Eqn 5.3.  
Removed most ties from TestKSTwoSamples testing data sets (by happenstance, subtracting `1/len(x)/len(y)/4.0` from the first data set happened to leave all the statistics and p-values the same, due to the original construction of the `x`, `y` pairs in the tests.)

pvanmulbregt commented:
The errors appear unrelated to the PR.

* `ci/circleci: pypy3` immediately failed with a http 404 error getting `http://security.debian.org/debian-security/pool/updates/main/o/openssl/libssl1.0.0_1.0.1t-1+deb8u10_amd64.deb` and never attended to build.  Previously also seen in gh-9900.

* `ci/circleci: build_docs` failed with `kpathsea:make_tex: Invalid fontname `[FreeSerif.otf]/OT', contains '['`. 
 I suspect that this is gh-9988.

pvanmulbregt commented:
Rebased and force-pushed to avoid gh-9988.

chrisb83 commented:
I had a quick look at the articles I mentioned above: 

10.1080/01621459.1976.10481562:
The calculation is also based on  Korolyuk using a recursion formula. The authors claim that "Hodges [3] did suggest a computational method for the one-sided test, but he computed no critical values, and his method appears less efficient than the procedure we now present. " Otherwise they do not provide any further insights on treatment of ties or an approximation.

10.1080/01621459.1969.10501082:
I did not see anything on treatment of ties. The author discusses ways to approximate the p-values (including a continuity correction) but I don't see any argument against Hodge's method.

So I think we should go ahead with the approximation of Hodge's which appears to work very well.

chrisb83 commented:
After the fix of the approximation, only open point is the treatment of ties. Any ideas? One simple option would be to raise a warning in that case and switch to "asymp" as in R.

pvanmulbregt commented:
Ties don't always affect the statistic. In Hodges' lattice, a tie means the path goes diagonal from (a,b) to (a+1, b+1). If the two paths  (a,b) to (a+1, b) to (a+1, b+1) and  (a,b) to (a, b+1) to (a+1, b+1) are both inside the strip then it doesn't matter. I.e. the statistic won't change of all the x data is shifted left (or right) by a small delta.
[More generally if a value is seen r times in one sample and t times in the other, the diagonal could be replaced with either of the two paths 
a,b -> a+r,b -> a+r,b+s
a,b -> a,b+s -> a+r,b+s]

If neither replacement path goes outside the strip, the exact computation is actually doing the right thing, so no need to switch to asymp. 


> On Apr 16, 2019, at 1:19 PM, Christoph Baumgarten <notifications@github.com> wrote:
> 
> After the fix of the approximation, only open point is the treatment of ties. Any ideas? One simple option would be to raise a warning in that case and switch to "asymp" as in R.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

chrisb83 commented:
OK, does it mean there is no specific treatment of ties is required? If that is the case, I think this PR is complete.

@rgommers this PR changes results. Can it be merged like this without an associated issue / bug? Can you advise please?

old signature: 
`ks_2samp(data1, data2)` would run two-sided test and return the approximated p-value

new signature:
`ks_2samp(data1, data2, alternative="two-sided", method="auto")` still runs the two-sided tests but returns the exact p-value for small samples and the approximated value for large samples. `method = "asymp"` would be equivalent to the old version but `auto` is the better choice.

rgommers commented:
> @rgommers this PR changes results. Can it be merged like this without an associated issue / bug? Can you advise please?

As far as I can tell from a quick read, the results change because they get *more correct* (like the length 1 example in the PR description, and the match with R). In that case, this can be considered a bug fix. It would be good to add a note in the release notes under backwards incompatible changes, but no separate issue seems needed.

pvanmulbregt commented:
gh-9786 is the Issue for small sample computation inaccuracies.

chrisb83 commented:
> As far as I can tell from a quick read, the results change because they get more correct

Yes, indeed.

I think we can consider to add this to 1.3.0

rgommers commented:
@chrisb83 go ahead and push the green button if you're happy:)

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5054,18 +5059,228 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+        minj = min(minj, n)
+        maxj = min(int(np.ceil((ng * i + h) / mg)), n + 1)
+        if maxj <= minj:
+            return 0
```
The `return 0` code path is never hit in tests

pvanmulbregt replied:
I was able to hit several of these paths by calling the private functions directly, and I've added some tests which execute those paths.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5054,18 +5059,228 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    # B is an array just holding a few values of B(x,y), the ones needed.
+    # B[j] == B(x_j, j)
+    if lxj == 0:
+        return np.round(special.binom(m + n, n))
```
This `return` code path is never hit in tests

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5054,18 +5059,228 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+            Bj -= dec
+        B[j] = Bj
+        if not np.isfinite(Bj):
+            raise FloatingPointError()
```
the raise is never hit in tests

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5054,18 +5059,228 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+        bin = np.round(special.binom((m-xj[j]) + (n - j), n-j))
+        term = B[j] * bin
+        if not np.isfinite(term):
+            raise FloatingPointError()
```
this raise is never hit in tests either

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
     n1 = data1.shape[0]
     n2 = data2.shape[0]
+    if min(n1, n2) == 0:
+        raise ValueError('Data passed to ks_2samp must not be empty')
```
`ValueError` not checked by tests

pvanmulbregt replied:
Now checked in TestKSTwoSamples::test_argument_checking.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
+            mode = 'asymp'
+    elif mode == 'exact':
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        if n1g >= np.iinfo(np.int).max / n2g:
```
This `if` is never true for tests

pvanmulbregt replied:
This one is a little time- and memory-consuming to trigger as it requires constructing two arrays each with about 3 billion elements.

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
+                        num_paths = _count_paths_outside_method(n1, n2, g, h)
+                        bin = special.binom(n1 + n2, n1)
+                        if not np.isfinite(bin) or not np.isfinite(num_paths) or num_paths > bin:
+                            raise FloatingPointError()
```
`raise` never hit by tests

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
+                            "Switching to mode=%s" % mode, RuntimeWarning)
+                else:
+                    if prob > 1 or prob < 0:
+                        mode = 'asymp'
```
control flow never reaches this branch in tests

pvanmulbregt replied:
It is protection against accumulation of floating point rounding errors taking the result out of the interval [0, 1], especially as `special.binom(a, b)` is not integral. But an example would be good.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5054,18 +5059,228 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+            # Set some carried-over elements to 0
+            A[maxj - minj:maxj - minj + (lastlen - curlen)] = 0
+        # Peel off one term from each of top and bottom of the binomial coefficient.
+        scaling_factor = i * 1.0 / (n + i)
```
The code was originally written in 2018 for Python2.x. In Python3 `1 / (n+i)` is sufficient.

pvanmulbregt commented:
ci/circleci:pypy3 faled immediately with 
```
ImportError: /home/circleci/project/pypy3.6/site-packages/numpy/core/_multiarray_umath.pypy3-70-x86_64-linux-gnu.so: undefined symbol: PyStructSequence_InitType2
```

tylerjereddy commented:
Yeah, this is looking great. Perhaps some tiny opportunities for polish but quite a bit of work with expert review and discussion of literature.

Thanks @pvanmulbregt & reviewers!

tylerjereddy commented:
I've added an improvement note and backwards incompatible change note to the [release notes wiki](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0), based on discussion above.

If you knew how little I know about stats, you might want to check those notes over :)
